spec/lib/food_robot_spec.rb

```diff
@@ -0,0 +1,294 @@
+      end
+
+      context 'with a non-functional schedule chain' do
+        let!(:chain) { create(:schedule_chain, day_of_week: monday.wday) }
+        let!(:recipient) { create(:recipient_schedule, schedule_chain: chain, position: 1) }
+        let!(:donor) { create(:donation_schedule, schedule_chain: chain, position: 2) }
+
+        subject { described_class.generate_log_entries(monday) }
+
+        it 'generates no logs' do
+          expect { subject }.not_to change { Log.count }
+        end
+      end
+
+      context 'with a weekly schedule chain' do
+        let!(:chain) { create(:schedule_chain, day_of_week: monday.wday, num_volunteers: 3) }
+        let!(:donor1) { create(:donation_schedule, schedule_chain: chain, position: 1) }
+        let!(:donor2) { create(:donation_schedule, schedule_chain: chain, position: 2) }
+        let!(:recipient) { create(:recipient_schedule, schedule_chain: chain, position: 3) }
+
+        context 'for a day other than the day of the schedule chain' do
+          subject { described_class.generate_log_entries(tuesday) }
+
+          it 'generates no logs' do
+            expect { subject }.not_to change { Log.count }
+          end
+        end
+
+        context 'for the day of the schedule chain' do
+          subject { described_class.generate_log_entries(monday) }
+
+          it 'generates a log for each donor' do
+            expect { subject }.to change { Log.count }.by(2)
+          end
+
+          it 'returns the number of created logs' do
+            n, _ = subject
+            expect(n).to eq(2)
+          end
+
+          it 'returns the number of skipped logs' do
+            _, n_skipped = subject
+            expect(n_skipped).to eq(0)
+          end
+
+          context 'with a log already generated' do
+            before do
+              create(
+                :log,
+                when: monday,
+                schedule_chain: chain,
+                donor: donor1.location,
+                recipients: [recipient.location]
+              )
+            end
+
+            it 'does not generate duplicate logs' do
+              expect { subject }.to change { Log.count }.by(1)
+            end
+
+            it 'returns the number of created logs' do
+              n, _ = subject
+              expect(n).to eq(1)
+            end
+
+            it 'returns the number of skipped logs' do
+              _, n_skipped = subject
+              expect(n_skipped).to eq(1)
+            end
+          end
+        end
+      end
+
+      context 'with a schedule chain ending at a hub' do
+        let!(:chain) { create(:schedule_chain, day_of_week: monday.wday, num_volunteers: 3) }
+        let!(:donor) { create(:donation_schedule, schedule_chain: chain, position: 1) }
+        let!(:hub) { create(:schedule, schedule_chain: chain, location: create(:hub), position: 2) }
+
+        subject { described_class.generate_log_entries(monday) }
+
+        it 'does not generate a log for the hub' do
```

Comment:
Got ya. I was thinking it might make sense to check that the log is for the donor as opposed to the hub, but since that's tested elsewhere we should be good. Thanks for the background!